File: frontend/src/types/book.ts
// src/types/book.ts

// バックエンドのスキーマに対応した型定義

// ✅ 書籍ステータスの列挙型
export enum BookStatusEnum {
  OWNED = "OWNED",
  WISHLIST = "WISHLIST",
}

export interface Book {
  id: number;
  title: string;
  volume: string;
  author: string;
  publisher: string;
  cover_image_url: string;
  published_date: string;
  user_id: number;
  status?: BookStatusEnum;
  is_favorite?: boolean;
  genres?: string[];
  isbn?: string;
}

export interface BookCreate {
  title: string;
  volume: string;
  author: string;
  publisher: string;
  cover_image_url: string;
  published_date: string;
  status?: BookStatusEnum;
  is_favorite?: boolean;
  genres?: string[];
  isbn?: string;
  // user_id: number; ✅ 削除：バックエンドで自動設定される
}

// ✅ 新規追加: 書籍更新用の型（任意のフィールドを更新可能）
export interface BookUpdate {
  title?: string;
  volume?: string;
  author?: string;
  publisher?: string;
  cover_image_url?: string;
  published_date?: string;
  status?: BookStatusEnum;
  is_favorite?: boolean;
  genres?: string[];
  // user_idは更新不可（セキュリティ上の理由）
}

export interface GoogleBookInfo {
  title: string;
  volume: string;
  authors: string[];
  publisher: string;
  cover_image_url: string;
  published_date: string;
}

export interface WishlistRegisterRequest {
  title: string;
  authors: string[];
  publisher: string;
  cover_image_url: string;
  published_date: string;
  isbn?: string;
  description?: string;
  categories?: string[];
  page_count?: number;
  language?: string;
}

// API関連の型
export interface ApiResponse<T> {
  data?: T;
  error?: string;
}

// UI状態の型
export interface UIState {
  isLoading: boolean;
  error: string | null;
}

// ✅ 新規追加: フィルター関連の型
export interface BookFilter {
  status?: BookStatusEnum;
  isFavorite?: boolean;
  searchQuery?: string;
  genres?: string[];
}

// ✅ 新規追加: 書籍一覧表示用の型
export interface BookListProps {
  books: Book[];
  filter?: BookFilter;
  onBookClick?: (book: Book) => void;
  onStatusChange?: (bookId: number, newStatus: BookStatusEnum) => void;
  onFavoriteToggle?: (bookId: number) => void;
}

File: backend/app/schemas/book.py
from datetime import date
from pydantic import BaseModel
from typing import Optional, List
from enum import Enum as PyEnum  # enum.Enum を別名で使用

# ✅ Enum 定義（str 継承でJSON互換）
class BookStatusEnum(str, PyEnum):
    OWNED = "owned"
    WISHLIST = "wishlist"
    NOT_OWNED = "not_owned"

# ✅ 共通の基本フィールド（空文字・空リストで一旦からOKな設計）
class BookBase(BaseModel):
    title: str = ""
    volume: str = ""
    author: str = ""
    publisher: Optional[str] = ""
    cover_image_url: Optional[str] = ""
    published_date: Optional[date] = None
    status: Optional[BookStatusEnum] = BookStatusEnum.OWNED
    is_favorite: bool = False
    isbn: Optional[str] = ""
    genres: List[str] = []

# ✅ 登録用（user_idはサーバー側で付与）
class BookCreate(BookBase):
    pass

# ✅ 更新用（すべてOptional）※ PATCH対応
class BookUpdate(BaseModel):
    title: Optional[str] = None
    volume: Optional[str] = None
    author: Optional[str] = None
    publisher: Optional[str] = None
    cover_image_url: Optional[str] = None
    published_date: Optional[date] = None
    status: Optional[BookStatusEnum] = None
    is_favorite: Optional[bool] = None
    genres: Optional[List[str]] = None
    isbn13: Optional[str] = None  # ← ISBN形式統一用の補助欄なら任意で

# schemas/book.py
class BookOut(BookBase):
    id: int
    user_id: int
    amazon_url: Optional[str] = None  # ✅ 追加

    class Config:
        orm_mode = True

# ✅ ISBNだけ受け取るAPIリクエスト用
class ISBNRequest(BaseModel):
    isbn: str
